README.md
# imu_knee_angle

code/history.txt
id	angle	time	type	imu_position
1	0	10	1	1
2	0	10	1	1
3	0	10	1	1
4	0	10	1	1
5	0	10	1	1
6	0	10	1	1
7	90	10	2	2
8	90	10	2	2
9	90	10	2	2
10	90	10	2	2
11	90	10	2	2
12	90	10	2	2
13	90	10	2	2
14	-1	10	4	3
15	-1	10	4	3
16	-1	10	4	3
17	-1	10	4	3
18	-1	10	4	3
19	-1	10	4	3
20	-1	10	4	3
21	90	10	4	3
22	70	10	4	3
23	-1	10	3	4
24	-1	10	3	4
25	-1	10	3	4
26	-1	10	3	4
27	90	10	3	4
28	90	10	3	4
29	90	10	3	4
30	60	10	3	4
31	60	10	3	4
32	60	10	3	4
33	30	10	1	5
34	30	10	1	5
35	30	10	1	5
36	30	10	1	5
37	30	10	1	5
38	30	10	1	5
39	30	10	1	5
40	30	10	1	5
41	30	10	1	5
42	30	10	1	5
43	30	10	1	5
44	45	10	1	6
45	45	10	1	6
46	45	10	1	6
47	45	10	1	6
48	45	10	1	6
49	45	10	1	6
50	45	10	1	6
51	45	10	1	6
52	45	10	1	6
53	45	10	1	6
54	45	10	1	6
55	45	10	1	6
56	60	10	1	7
57	60	10	1	7
58	60	10	1	7
59	60	10	1	7
60	60	10	1	7
61	60	10	1	7
62	60	10	1	7
63	60	10	1	7

code/trans.py
'''
代码作用
1. 做数据差值，使用linear方法，固定采样频率
2. 根据角度计算角速度
3. 生成角度计算的输入数据
'''

from scipy import interpolate
import math
import time
import numpy as np
import os

from config import AngelConfing

def MAD(dataset,n):
    median = np.median(dataset)
    deviations = abs(dataset - median)
    mad = np.median(deviations)
    remove_idx = np.where(abs(dataset - median) >  n * mad)
    cur=[]
    for item in list(remove_idx[0]):
        cur.append(dataset[item])
    print(cur)
    return list(remove_idx[0])

def three_sigma(dataset, n= 3):
    mean = np.mean(dataset)
    sigma = np.std(dataset)
    remove_idx = np.where(abs(dataset - mean) > n * sigma)
    cur=[]
    for item in list(remove_idx[0]):
        cur.append(dataset[item])
    print(cur)
    return list(remove_idx[0])

def get_data(path):
    with open(path) as f:
        data=eval(f.read())
    return data

def store(data,path,info):
    with open(path,'w') as f:
        f.write("%s\n"%str(time.time()))
        f.write("%s\n"%str(info))
        for item in data:
            f.write("%f\t%f\t%f\t%f\t%f\t%f\n"%\
                (item[1],item[2],item[3],item[4],item[5],item[6]))

def removeOutliers(remove_idx,data):
    res=[]
    for i in range(len(data)):
        if i not in remove_idx:
            res.append(data[i])
    return res

def get_inter(data,begin,end,diff):
    tot=begin
    whole=[] #差值后的数据
    nx=[] #新的 x
    while tot<end:
        whole.append([tot])
        nx.append(tot)
        tot+=diff

    x=[] #旧的 x
    for item in data:
        x.append(item[0])

    for i in range(len(data[0][1])):
        #对于每个数据点,做线性差值
        cur=[]
        for j in range(len(data)):
            cur.append(data[j][1][i])
        f=interpolate.interp1d(x,cur,kind='linear')
        y=f(nx)
        for j in range(len(whole)):
            whole[j].append(y[j])

    return whole

    #去掉离群点
    '''
    convert=list(map(list,zip(*res)))
    remove_idx=[]
    for i in range(4,6):
        print(i)
        
        #cur=MAD(convert[i],3)
        cur=three_sigma(convert[i],2)
        print(cur)
        print('\n')
        remove_idx+=cur

    res=removeOutliers(remove_idx,res)
    
    for i in range(len(res)):
        for j in range(4,7):
            if res[i][j]>100:
                res[i][j]=100
            if res[i][j]<-100:
                res[i][j]=-100
    '''
    
def trans(path='',diff=0.0):
    root_path='../data/'
    if diff==0.0: 
        # diff可以自己指定，如果没指定，就使用默认的
        diff=AngelConfing.diff
    if len(path)==0:
        # 如果不指定，默认将'../data' 下的所有文件转换到新的文件夹下
        all_path=[]
        dirs=os.listdir(root_path)
        if '.DS_Store' in dirs:
            dirs.remove('.DS_Store')
    else:
        all_path=[path]

    for item in all_path:
        cur_data=get_data(root_path+item)
        begin=math.ceil(data[0][0])
        end=math.floor(data[-1][0])
        cur_data=get_inter(cur_data,begin,end,diff)
        cur_path=item.replace('data','newdata',1)

        store(cur_data,cur_path,[begin,end,diff])
        # data.append(get_data(root_path + item))


    ##begin 和 end 不如放到 JointAngel去判断
    #### 记得去改一下，或者再写一个外包函数去做判断，在JointAngel的基础上
   

if __name__=='__main__':
    # path=['../data/gj1.txt','../data/gj2.txt']
    trans()


code/PyImuSerial.py
import serial
import struct
import time

##两个imu分开写
state1 = 0
payload1 = list()

state2 = 0
payload2 = list()

# +SPPDATA=1,028,
header1 = bytes([0x2B, 0x53, 0x50, 0x50, 0x44, 0x41, 0x54, \
    0x41, 0x3D, 0x31, 0x2C, 0x30, 0x32, 0x38, 0x2C])
# +SPPDATA=2,028,
header2= bytes([0x2B, 0x53, 0x50, 0x50, 0x44, 0x41, 0x54,\
 0x41, 0x3D, 0x32, 0x2C, 0x30, 0x32, 0x38, 0x2C])

class DataFrame():
    def __init__(self, buffer):
        self.device_addr = buffer[0]
        self.data = buffer[1:25]
        self.invariant = buffer[25:28]
        self.reserved = buffer[28:32]

    def __repr__(self):
        vals = struct.unpack("<6f", self.data)
        vals_str = "{:8.4f},{:8.4f},{:8.4f},{:8.4f},{:8.4f},{:8.4f}"\
        .format(vals[0], vals[1], vals[2], vals[3], vals[4], vals[5])
        return vals_str

def getframe1(b):
    global state1
    global payload1

    b = int.from_bytes(b, byteorder='big')
    if state1 == -1 and b == 0x2B:
        state1 = 0
        return False
    elif state1 >= 0 and state1 < 14:
        if header1[state1 + 1] == b:
            state1 = state1 + 1
        else:
            state1 = -1
        return False
    elif state1 >= 14 and state1 < 45:
        payload1.append(b)
        state1 = state1 + 1
        return False
    elif state1 == 45:
        payload1.append(b)
        state1 = -1
        return True

def getframe2(b):
    global state2
    global payload2

    b = int.from_bytes(b, byteorder='big')
    if state2 == -1 and b == 0x2B:
        state2 = 0
        return False
    elif state2 >= 0 and state2 < 14:
        if header2[state2 + 1] == b:
            state2 = state2 + 1
        else:
            state2 = -1
        return False
    elif state2 >= 14 and state2 < 45:
        payload2.append(b)
        state2 = state2 + 1
        return False
    elif state2 == 45:
        payload2.append(b)
        state2 = -1
        return True

def getval(cur):
    whole=[]
    res=cur.split(',')
    for item in res:
        whole.append(eval(item))
    return whole

def store(data,path):
    with open(path,'w') as f:
        for item in data:
            f.write("%f\t%f\t%f\t%f\t%f\t%f\t%f\n"%\
                (item[0],item[1][0],item[1][1],item[1][2],item[1][3],item[1][4],item[1][5]))


def data_collection(second=30,name='imu'):
    global state1
    global payload1
    global state2
    global payload2
    port_name = "/dev/tty.usbserial-A50285BI"
    port_bps = 921600
    serial_port = serial.Serial(port_name, port_bps, timeout=1)
    print(serial_port)
    serial_port.write("AT+SPPSEND=1,6,view=1\r\n".encode(encoding="ascii"))
    time.sleep(3)
    serial_port.write("AT+SPPSEND=2,6,view=1\r\n".encode(encoding="ascii"))

    print("开始收集数据")
    begin=time.time()

    path1='../data/%d_1.txt'%(name)
    path2='../data/%d_2.txt'%(name)
    imu1=[]
    imu2=[]

    while True:
        data=serial_port.read()
        cur=time.time()
        if cur-begin>second:
            print("数据收集完毕")
            break
        if getframe1(data):
            frame = DataFrame(bytes(payload1))
            frame = getval(str(frame))
            imu1.append((cur,frame))
            payload1 = []
        if getframe2(data):
            frame = DataFrame(bytes(payload2))
            frame = getval(str(frame))
            imu2.append((cur,frame))
            payload2 = []

    store(imu1,path1)
    store(imu2,path2)


if __name__=='__main__':
    data_collection(10,name='90_1')


code/JointAngel.py
import numpy as np
from scipy import interpolate
import datetime
import math

from config import AngelConfing

def get_time(timestr):
	timestr=timestr.strip()
	cur=datetime.datetime.strptime(timestr,'%Y-%m-%d %H:%M:%S.%f')
	return cur.timestamp()

def get_raw_data(path='../data/1.csv'):
	data=[]
	#[a1,a2,a3,g1,g2,g3]
	with open(path) as f:
		context=f.read().split('\n')
		whole=[]
		for item in context:
			cur=item.split('\t')
			whole.append(cur)
		names=whole[0]
		whole=whole[1:]
		device_name=[]
		for item in whole:
			if item[0] not in device_name and len(item[0]):
				device_name.append(item[0])
		
		for i in range(len(device_name)):
			data.append([])

		for item in whole:
			if len(item)==0 or len(item[0])==0:
				continue
			cur_device=device_name.index(item[0])
			cur=[]
			cur.append(get_time(item[3]))
			for i in range(4,7):
				cur.append(float(item[i]))
			for i in range(7,10):
				cur.append(float(item[i]))

			data[cur_device].append(cur)
	#计算[g1',g2',g3'],需要先做差值，再计算

	return data

def get_inter(data,begin,end,diff):
	tot=begin
	whole=[]
	nx=[]
	while tot<end:
		whole.append([tot])
		nx.append(tot)
		tot=float('%.6f'%(tot+diff))

	x=[]
	for item in data:
		x.append(item[0])

	for i in range(1,len(data[0])):
		cur=[]
		for j in range(len(data)):
			cur.append(data[j][i])
		f=interpolate.interp1d(x,cur,kind='linear')
		y=f(nx)
		for j in range(len(whole)):
			#whole[j].append(y[j])
			whole[j].append(float(y[j]))

	return whole

def get_angular_acceleration(data,diff,index):
	vel_dot=[]
	for i in range(len(data)):
		cur=[]
		for j in range(len(index)):
			res=0.0
			if i >1 and i<len(data)-2:
				res=(data[i-2][index[j]]-8*data[i-1][index[j]]+8*data[i+1][index[j]]-data[i+2][index[j]])/(12*diff)
			elif i==1 or i==len(data)-2:
				res=(data[i+1][index[j]]-data[i-1][index[j]])/(2*diff)
			elif i==0:
				res=(data[i+1][index[j]]-data[i][index[j]])/diff
			else:
				res=(data[i][index[j]]-data[i-1][index[j]])/diff
			res=float('%.6f'%(res))
			cur.append(res)
		vel_dot.append(cur)
	for i in range(len(data)):
		data[i]+=vel_dot[i]

	return data

def c(tt):
	return math.cos(tt)

def s(tt):
	return math.sin(tt)

def sqrt(tt):
	return math.sqrt(tt)

def pow(tt):
	return math.pow(tt,2)

def get_pos(input_data,params,output):
	o1x=params[0][0]
	o1y=params[1][0]
	o1z=params[2][0]
	o2x=params[3][0]
	o2y=params[4][0]
	o2z=params[5][0]

	m=input_data.shape[0]
	for i in range(m):
		acc_joint1_x = input_data[i][4] * ( input_data[i][3] * o1y - input_data[i][4] * o1x ) - input_data[i][5] * ( input_data[i][5] * o1x - input_data[i][3] * o1z ) + ( input_data[i][7] * o1z - input_data[i][8] * o1y )
		acc_joint1_y = input_data[i][5] * ( input_data[i][4] * o1z - input_data[i][5] * o1y ) - input_data[i][3] * ( input_data[i][3] * o1y - input_data[i][4] * o1x ) + ( input_data[i][8] * o1x - input_data[i][6] * o1z )
		acc_joint1_z = input_data[i][3] * ( input_data[i][5] * o1x - input_data[i][3] * o1z ) - input_data[i][4] * ( input_data[i][4] * o1z - input_data[i][5] * o1y ) + ( input_data[i][6] * o1y - input_data[i][7] * o1x )

		acc_joint2_x = input_data[i][13] * ( input_data[i][12] * o2y - input_data[i][13] * o2x ) - input_data[i][14] * ( input_data[i][14] * o2x - input_data[i][12] * o2z ) + ( input_data[i][16] * o2z - input_data[i][17] * o2y )
		acc_joint2_y = input_data[i][14] * ( input_data[i][13] * o2z - input_data[i][14] * o2y ) - input_data[i][12] * ( input_data[i][12] * o2y - input_data[i][13] * o2x ) + ( input_data[i][17] * o2x - input_data[i][15] * o2z )
		acc_joint2_z = input_data[i][12] * ( input_data[i][14] * o2x - input_data[i][12] * o2z ) - input_data[i][13] * ( input_data[i][13] * o2z - input_data[i][14] * o2y ) + ( input_data[i][15] * o2y - input_data[i][16] * o2x )

		output[i][0] =  sqrt(\
			pow( input_data[i][0] - acc_joint1_x ) + pow( input_data[i][1] - acc_joint1_y ) + pow( input_data[i][2] - acc_joint1_z )) - \
		sqrt(\
			pow( input_data[i][9] - acc_joint2_x ) + pow( input_data[i][10] - acc_joint2_y) + pow( input_data[i][11] - acc_joint2_z ))

	return output


def get_axis(input_data,params,output):
	theta_1=params[0][0]
	theta_2=params[1][0]
	phi_1=params[2][0]
	phi_2=params[3][0]

	m=input_data.shape[0]
	for i in range(m):
		output[i][0]=sqrt(\
			pow(input_data[i][1]*s(theta_1)- input_data[i][2]*c(phi_1)*s(theta_1) )+\
			pow(input_data[i][2]*c(phi_1)*c(theta_1)- input_data[i][0]*s(phi_1) )+\
			pow(input_data[i][0]*c(phi_1)*s(theta_1)- input_data[i][1]*c(phi_1)*c(theta_1) ))-\
		sqrt(\
			pow(input_data[i][4]*s(theta_2)- input_data[i][5]*c(phi_2)*s(theta_2) )+\
			pow(input_data[i][5]*c(phi_2)*c(theta_2)- input_data[i][3]*s(phi_2) )+\
			pow(input_data[i][3]*c(phi_2)*s(theta_2)- input_data[i][4]*c(phi_2)*c(theta_2) ))

	return output

def get_angel_acc(j1,j2,a1,a2,g1,g2,g_dot1,g_dot2,o1,o2):
	c=np.zeros((3,1))
	x1=np.zeros((3,1))
	x2=np.zeros((3,1))
	y1=np.zeros((3,1))
	y2=np.zeros((3,1))

	acc1=np.zeros((2,1))
	acc2=np.zeros((2,1))

	a1_dot=np.zeros((1,3))
	a2_dot=np.zeros((1,3))

	p1=0.0
	p2=0.0
	q1=0.0
	q2=0.0
	angle_acc=0.0

	c=np.array([[1.0],[2.0],[3.0]])

	p1=np.cross(g1,o1.transpose())
	p2=np.cross(g1,p1)
	p3=np.cross(g_dot1,o1.transpose())
	a1_dot=a1-(p2+p3)

	p_1=np.cross(g2,o2.transpose())
	p_2=np.cross(g2,p_1)
	p_3=np.cross(g_dot2,o2.transpose())
	a2_dot=a2-(p_2+p_3)

	x1=np.cross(j1.transpose(),c.transpose())
	y1=np.cross(j1.transpose(),x1)
	x2=np.cross(j2.transpose(),c.transpose())
	y2=np.cross(j2.transpose(),x2)

	p1=np.dot(a1_dot,x1.transpose())[0][0]
	p2=np.dot(a1_dot,y1.transpose())[0][0]
	q1=np.dot(a2_dot,x2.transpose())[0][0]
	q2=np.dot(a2_dot,y2.transpose())[0][0]

	acc1=np.array([[p1],[p2]])
	acc2=np.array([[q1],[q2]])

	angle_acc=math.acos(np.dot(acc1.transpose(),acc2)[0][0]\
		/(np.linalg.norm(acc1)*np.linalg.norm(acc2)))
	return math.degrees(angle_acc)

class joint_angel():
	def __init__(self,data,test_data):
		self.DATASET_NUM=len(data[0])
		self.DELTA_T=AngelConfing.diff
		self.ITER_STEP=AngelConfing.ITER_STEP
		self.ITER_CNT=AngelConfing.ITER_CNT
		self.params_axis=np.zeros((4,1))
		self.params_pos=np.zeros((6,1))
		self.vj1=np.zeros((3,1))
		self.vj2=np.zeros((3,1))
		self.o1=np.zeros((3,1))
		self.o2=np.zeros((3,1))
		self.sname=''
		self.imu_data_1=np.array(data[0])
		self.imu_data_2=np.array(data[1])

		self.imu_test_data_1=np.array(test_data[0])
		self.imu_test_data_2=np.array(test_data[1])
		self.TEST_DATASET_NUM=len(test_data[0])

	def get_jacobian(self,func,input_data,params,output):
		m=input_data.shape[0]
		n=params.shape[0]

		out0=np.zeros((m,1))
		out1=np.zeros((m,1))
		param0=np.zeros((n,1))
		param1=np.zeros((n,1))

		for j in range(n):
			param0=params.copy()
			param1=params.copy()
			param0[j][0]-=self.ITER_STEP
			param1[j][0]+=self.ITER_STEP
			
			out1 = func(input_data,param1,out1)
			out0 = func(input_data,param0,out0)

			output[0:m,j:j+1]=(out1-out0)/(2*self.ITER_STEP)

		return output

	def gauss_newton(self,func,input_data,output,params):


		m=input_data.shape[0]
		n=params.shape[0]

		#jacobian
		jmat=np.zeros((m,n))
		r=np.zeros((m,1))
		tmp=np.zeros((m,1))

		pre_mse=0.0
		mse=0.0

		result=[]

		for i in range(self.ITER_CNT):			
			mse=0.0
			tmp=func(input_data,params,tmp)
			r=output-tmp
			jmat=self.get_jacobian(func,input_data,params,jmat)

			mse=np.dot(r.transpose(),r)
			mse/=m
			if abs(mse- pre_mse )<1e-6:
				print("均方根误差更新过小，已收敛")
				break

			pre_mse=mse
			a=np.dot(jmat.transpose(),jmat)

			b=np.linalg.pinv(a)
			c=np.dot(b,jmat.transpose())
			d=np.dot(c,r)
			result.append((mse,params.copy()))

			print("i=%d, mes:%lf"%(i,mse))

			params+=d

		pre=params
		minn=10000.0
		for item in result:
			if item[0]<minn:
				pre=item[1]
				minn=item[0]

		return pre

	def joint_pos(self):
		input_data=np.zeros((self.DATASET_NUM,18))
		output=np.zeros((self.DATASET_NUM,1))

		for i in range(self.DATASET_NUM):
			k=0
			for j in range(9):
				input_data[i][k]=self.imu_data_1[i][j]
				k+=1
			for j in range(9):
				input_data[i][k]=self.imu_data_2[i][j]
				k+=1
			output[i][0]=0.0

		for i in range(6):
			self.params_pos[i][0]=0.1

		self.params_pos=self.gauss_newton(get_pos,input_data,output,self.params_pos)

		self.o1=np.array([[self.params_pos[0][0]],[self.params_pos[1][0]],[self.params_pos[2][0]]])
		self.o2=np.array([[self.params_pos[3][0]],[self.params_pos[4][0]],[self.params_pos[5][0]]])

		print(self.o1,self.o2)

	def joint_axis(self):
		#计算关节轴向数据输入接口
		input_data=np.zeros((self.DATASET_NUM,6))
		output=np.zeros((self.DATASET_NUM,1))

		for i in range(self.DATASET_NUM):
			k=0
			for j in range(3,6):
				input_data[i][k]=self.imu_data_1[i][j]
				k+=1
			for j in range(3,6):
				input_data[i][k]=self.imu_data_2[i][j]
				k+=1
			output[i][0]=0.0

		for i in range(4):
			self.params_axis[i][0]=0.5

		self.params_axis=self.gauss_newton(get_axis,input_data,output,self.params_axis)

		self.vj1=np.array([[c(self.params_axis[2][0])*c(self.params_axis[0][0])],\
			[c(self.params_axis[2][0]*s(self.params_axis[0][0]))],\
			[s(self.params_axis[2][0])]])
		self.vj2=np.array([[c(self.params_axis[3][0])*c(self.params_axis[1][0])],\
			[c(self.params_axis[3][0])*s(self.params_axis[1][0])],\
			[s(self.params_axis[3][0])]])
		##做一个符号匹配
		flag=False
		if self.vj1[0][0]*self.vj2[0][0]<1e-10:
			flag=True
		if flag:
			self.vj1=-self.vj1
		print('j1,j2:')
		print(self.vj1)
		print(self.vj2)

	def test_angel(self):
		angle_acc = 0.0
		angle_gyr = 0.0
		angle_acc_gyr = 0.0
		prev_angle_gyr=0.0
		prev_angle_acc_gyr=0.0

		SUM=0.0
		cnt=0
		LAMBDA=0.01

		whole=[]

		for i in range(self.TEST_DATASET_NUM):
			cnt+=1
			a1=np.array([[self.imu_test_data_1[i][0],self.imu_test_data_1[i][1],self.imu_test_data_1[i][2]]])
			a2=np.array([[self.imu_test_data_2[i][0],self.imu_test_data_2[i][1],self.imu_test_data_2[i][2]]])
			g1=np.array([[self.imu_test_data_1[i][3],self.imu_test_data_1[i][4],self.imu_test_data_1[i][5]]])
			g2=np.array([[self.imu_test_data_2[i][3],self.imu_test_data_2[i][4],self.imu_test_data_2[i][5]]])
			g_dot1=np.array([[self.imu_test_data_1[i][6],self.imu_test_data_1[i][7],self.imu_test_data_1[i][8]]])
			g_dot2=np.array([[self.imu_test_data_2[i][6],self.imu_test_data_2[i][7],self.imu_test_data_2[i][8]]])

			angle_acc=get_angel_acc(self.vj1,self.vj2,a1,a2,g1,g2,g_dot1,g_dot2,self.o1,self.o2)

			
			aaa=np.dot(g1,self.vj1)-np.dot(g2,self.vj2)
			SUM=SUM+aaa[0][0]

			if cnt>3:
				cur={}
				angle_gyr=SUM*self.DELTA_T
				angle_acc_gyr=LAMBDA*angle_acc+(1- LAMBDA )*(prev_angle_acc_gyr+angle_gyr-prev_angle_gyr )
				print("angle_acc_gyr:",angle_acc_gyr)
				print("angle_acc:",angle_acc)
				# print("angle_gyr",angle_gyr)
				cur["angle_acc_gyr"]=angle_acc_gyr
				cur['angle_acc']=angle_acc
				cur["angle_gyr"]=angle_gyr
				whole.append(cur)
			
				
			if cnt<=3:
				prev_angle_acc_gyr=angle_acc
				prev_angle_gyr=angle_gyr
			else:
				prev_angle_acc_gyr=angle_acc_gyr
				prev_angle_gyr=angle_gyr

		with open(self.sname,'w') as f:
			f.write(str(whole))

def get_data(path):
	whole=[]
	with open(path) as f:
		context=f.read()
		context=context.split('\n')
		for item in context:
			if len(item)==0:
				continue
			cur=[]
			item=item.split('\t')
			for i in item:
				cur.append(float(i))
			whole.append(cur)
	return whole

def main(train,test):

	p1='../data/%d_1.txt'%(train)
	p2='../data/%d_2.txt'%(train)
	p3='../data/%d_1.txt'%(test)
	p4='../data/%d_2.txt'%(test)

	d1=get_data(p1)
	d2=get_data(p2)
	lens1=min(len(d1),len(d2))

	d1=d1[:lens1]
	d2=d2[:lens1]
	train_data=[d1,d2]

	d3=get_data(p3)
	d4=get_data(p4)
	lens2=min(len(d3),len(d4))

	d3=d3[:lens2]
	d4=d4[:lens2]
	test_data=[d3,d4]

	#去掉时间戳
	for i in range(len(train_data)):
		for j in range(len(train_data[i])):
			train_data[i][j]=train_data[i][j][1:]

	for i in range(len(test_data)):
		for j in range(len(test_data[i])):
			test_data[i][j]=test_data[i][j][1:]

	for i in range(len(train_data)):
		train_data[i]=get_angular_acceleration(train_data[i],AngelConfing.diff,[3,4,5])

	for i in range(len(test_data)):
		test_data[i]=get_angular_acceleration(test_data[i],AngelConfing.diff,[3,4,5])


	a=joint_angel(train_data,test_data)
	a.joint_axis()
	a.joint_pos()

	t1=np.dot(a.o1.transpose(),a.vj1)[0][0]
	t2=np.dot(a.o2.transpose(),a.vj2)[0][0]
	tt=(t1+t2)/2
	a.o1-=a.vj1*tt
	a.o2-=a.vj2*tt
	a.sname='../result/%d_%d.txt'%(train,test)
	a.test_angel()

if __name__ == '__main__':
	main(57,56)
